File: Scotch/HttpInteraction.fs
namespace Scotch

open System
open System.Text
open System.Net
open System.Net.Http
open System.Collections.Generic

type HeaderDictionary = IDictionary<string, string>

type Request =
    { Method : string
      URI : string
      RequestHeaders : HeaderDictionary
      ContentHeaders : HeaderDictionary
      Body : string }

type Status =
    { Code : HttpStatusCode
      Message : string }

type Response =
    { Status : Status
      ResponseHeaders : HeaderDictionary
      ContentHeaders : HeaderDictionary
      Body : string
      HttpVersion : Version } with
    member this.ToHttpResponseMessage () =
        let result = new HttpResponseMessage(this.Status.Code)
        result.ReasonPhrase <- this.Status.Message
        result.Version <- this.HttpVersion
        for h in this.ResponseHeaders do result.Headers.TryAddWithoutValidation(h.Key, h.Value.ToString()) |> ignore
        let content = new ByteArrayContent(Encoding.UTF8.GetBytes(this.Body))
        for h in this.ContentHeaders do content.Headers.TryAddWithoutValidation(h.Key, h.Value.ToString()) |> ignore
        result.Content <- content
        result

type HttpInteraction =
    { Request : Request
      Response : Response
      RecordedAt : DateTimeOffset }

module Helpers =
    let toHeaders (headers: Headers.HttpHeaders) =
        headers
        |> Seq.map (fun h -> (h.Key, String.Join(",", h.Value)))
        |> dict

    let toStringAsync (content:HttpContent) =
        match content with
        | null -> async {return ""}
        | _ -> content.ReadAsStringAsync() |> Async.AwaitTask

    let toContentHeaders (content:HttpContent) =
        match content with
        | null -> [||] |> dict
        | _ -> toHeaders content.Headers

    let toRequestAsync (request: HttpRequestMessage) =
        async {
            let! requestBody = toStringAsync request.Content
            return
                {Method = request.Method.ToString()
                 URI = request.RequestUri.ToString()
                 RequestHeaders = toHeaders request.Headers
                 ContentHeaders = toContentHeaders request.Content
                 Body = requestBody
                 }
        }

    let toResponseAsync (response: HttpResponseMessage) =
        async {
            let! responseBody = toStringAsync response.Content
            return
                {Status = {Code = response.StatusCode; Message = response.ReasonPhrase}
                 ResponseHeaders = toHeaders response.Headers
                 ContentHeaders = toContentHeaders response.Content
                 Body = responseBody
                 HttpVersion = response.Version}
        }

    let requestsMatch receivedRequest recordedRequest =
        receivedRequest.Method.Equals(recordedRequest.Method, StringComparison.InvariantCultureIgnoreCase)
        && receivedRequest.URI.Equals(recordedRequest.URI, StringComparison.InvariantCultureIgnoreCase)

File: Scotch/HttpInteraction.cs
using System.Net;
using System.Net.Http.Headers;
using System.Text;

namespace Scotch;

public class Request
{
    public string? Body { get; set; }
    public IDictionary<string, string>? ContentHeaders { get; set; }
    public string Method { get; set; }
    public IDictionary<string, string> RequestHeaders { get; set; }
    public string Uri { get; set; }
}

public class Status
{
    public HttpStatusCode Code { get; set; }
    public string? Message { get; set; }
}

public class Response
{
    public string? Body { get; set; }
    public IDictionary<string, string>? ContentHeaders { get; set; }

    public Version HttpVersion { get; set; }
    public IDictionary<string, string>? ResponseHeaders { get; set; }
    public Status Status { get; set; }

    public HttpResponseMessage ToHttpResponseMessage(HttpRequestMessage requestMessage)
    {
        var result = new HttpResponseMessage(Status.Code);
        result.ReasonPhrase = Status.Message;
        result.Version = HttpVersion;
        foreach (var h in ResponseHeaders ?? new Dictionary<string, string>()) result.Headers.TryAddWithoutValidation(h.Key, h.Value.ToString());

        var content = new ByteArrayContent(Encoding.UTF8.GetBytes(Body ?? string.Empty));
        foreach (var h in ContentHeaders ?? new Dictionary<string, string>()) content.Headers.TryAddWithoutValidation(h.Key, h.Value.ToString());

        result.Content = content;
        result.RequestMessage = requestMessage;
        return result;
    }
}

public class HttpInteraction
{
    public DateTimeOffset RecordedAt { get; set; }
    public Request Request { get; set; }
    public Response Response { get; set; }
}

public static class Helpers
{
    public static bool RequestsMatch(Request receivedRequest, Request recordedRequest)
    {
        return receivedRequest.Method.Equals(recordedRequest.Method, StringComparison.OrdinalIgnoreCase)
               && receivedRequest.Uri.Equals(recordedRequest.Uri, StringComparison.OrdinalIgnoreCase);
    }

    public static async Task<Request> ToRequestAsync(HttpRequestMessage request, List<string>? headersToHide = null)
    {
        var requestBody = await ToStringAsync(request.Content);
        return new Request
        {
            Method = request.Method.ToString(),
            Uri = request.RequestUri.ToString(),
            RequestHeaders = ToHeaders(request.Headers, headersToHide),
            ContentHeaders = ToContentHeaders(request.Content),
            Body = requestBody
        };
    }

    public static async Task<Response> ToResponseAsync(HttpResponseMessage response, List<string>? headersToHide = null)
    {
        var responseBody = await ToStringAsync(response.Content);
        return new Response
        {
            Status = new Status
            {
                Code = response.StatusCode,
                Message = response.ReasonPhrase
            },
            ResponseHeaders = ToHeaders(response.Headers, headersToHide),
            ContentHeaders = ToContentHeaders(response.Content),
            Body = responseBody,
            HttpVersion = response.Version
        };
    }

    private static IDictionary<string, string> ToContentHeaders(HttpContent? content)
    {
        return content == null ? new Dictionary<string, string>() : ToHeaders(content.Headers);
    }

    private static IDictionary<string, string> ToHeaders(HttpHeaders headers, List<string>? headersToHide = null)
    {
        headersToHide ??= new List<string>();

        IDictionary<string, string> dict = new Dictionary<string, string>();
        foreach (var h in headers) dict.Add(h.Key, headersToHide.Contains(h.Key) ? "********" : string.Join(",", h.Value));

        return dict;
    }

    private static async Task<string> ToStringAsync(HttpContent? content)
    {
        return content == null ? string.Empty : await content.ReadAsStringAsync();
    }
}
